--- prev/cms.c	2018-08-14 22:48:56.000000000 +1000
+++ apps/cms.c	2018-09-25 10:59:03.207170134 +1000
@@ -109,6 +109,7 @@
 };
 
 int MAIN(int, char **);
+static X509_CRL *load_crl_for_cms(char *infile, int format);
 
 int MAIN(int argc, char **argv)
 {
@@ -119,6 +120,7 @@
     const char *inmode = "r", *outmode = "w";
     char *infile = NULL, *outfile = NULL, *rctfile = NULL;
     char *signerfile = NULL, *recipfile = NULL;
+    char *crlfile = NULL;
     STACK_OF(OPENSSL_STRING) *sksigners = NULL, *skkeys = NULL;
     char *certfile = NULL, *keyfile = NULL, *contfile = NULL;
     char *certsoutfile = NULL;
@@ -483,6 +485,10 @@
             if (!args[1])
                 goto argerr;
             certfile = *++args;
+        } else if (!strcmp(*args, "-CRLfile")) {
+            if (!args[1])
+                goto argerr;
+            crlfile = *++args;
         } else if (!strcmp(*args, "-CAfile")) {
             if (!args[1])
                 goto argerr;
@@ -966,6 +972,13 @@
                     goto end;
                 }
             }
+            if (crlfile) {
+                X509_CRL *crl = load_crl_for_cms(crlfile, FORMAT_PEM);
+                if (!crl) {
+                    exit(1);
+                }
+                CMS_add1_crl(cms, crl);
+            }
         } else
             flags |= CMS_REUSE_DIGEST;
         for (i = 0; i < sk_OPENSSL_STRING_num(sksigners); i++) {
@@ -1355,4 +1368,42 @@
     return 1;
 }
 
+static X509_CRL *load_crl_for_cms(char *infile, int format)
+{
+    X509_CRL *x = NULL;
+    BIO *in = NULL;
+
+    in = BIO_new(BIO_s_file());
+    if (in == NULL) {
+        ERR_print_errors(bio_err);
+        goto end;
+    }
+
+    if (infile == NULL)
+        BIO_set_fp(in, stdin, BIO_NOCLOSE);
+    else {
+        if (BIO_read_filename(in, infile) <= 0) {
+            perror(infile);
+            goto end;
+        }
+    }
+    if (format == FORMAT_ASN1)
+        x = d2i_X509_CRL_bio(in, NULL);
+    else if (format == FORMAT_PEM)
+        x = PEM_read_bio_X509_CRL(in, NULL, NULL, NULL);
+    else {
+        BIO_printf(bio_err, "bad input format specified for input crl\n");
+        goto end;
+    }
+    if (x == NULL) {
+        BIO_printf(bio_err, "unable to load CRL\n");
+        ERR_print_errors(bio_err);
+        goto end;
+    }
+
+ end:
+    BIO_free(in);
+    return (x);
+}
+
 #endif
